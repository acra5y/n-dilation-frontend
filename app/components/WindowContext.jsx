import React, { useContext } from "react";

const WindowContext = React.createContext();

export const useWindowContext = () => useContext(WindowContext);

export default WindowContext;
