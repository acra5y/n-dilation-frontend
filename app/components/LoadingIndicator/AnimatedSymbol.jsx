import React, { useEffect, useState } from "react";
import styled from "styled-components";

import { ANIMATION_TIME, INTERVAL, MINIMUM_TIME_VISIBLE } from "./constants";
import getRandomInt from "../../lib/getRandomInt";
import { useWindowContext } from "../WindowContext";

const StyledSymbol = styled.span.attrs(props => ({
    style: {
        flexBasis: `${props.basis}%`,
        margin: `calc(4rem * ${props.offsetTop}) 3px 3px`,
        opacity: props.opacity,
    },
}))`
    white-space: nowrap;
    transition: opacity ${ANIMATION_TIME}ms ease-out;
`;
StyledSymbol.displayName = "StyledSymbol";

const AnimatedSymbol = props => {
    const [opacity, setOpacity] = useState(0);

    const window = useWindowContext();

    useEffect(() => {
        const { setTimeout, clearTimeout } = window;
        let i, j;

        const clearance = 100;

        const timeBeforeFadeIn = props.getRandomInt(clearance);
        const timeVisible = Math.max(
            MINIMUM_TIME_VISIBLE,
            props.getRandomInt(INTERVAL - timeBeforeFadeIn - ANIMATION_TIME)
        );

        i = setTimeout(() => {
            j = setTimeout(() => setOpacity(0), timeVisible);
            setOpacity(1);
        }, timeBeforeFadeIn);

        return () => {
            clearTimeout(i);
            clearTimeout(j);
        };
    }, [props.children]);

    return (
        <StyledSymbol
            basis={props.basis}
            offsetTop={props.offsetTop}
            opacity={opacity}
        >
            {props.children}
        </StyledSymbol>
    );
};

AnimatedSymbol.displayName = "AnimatedSymbol";

AnimatedSymbol.defaultProps = {
    getRandomInt,
};

export default AnimatedSymbol;
