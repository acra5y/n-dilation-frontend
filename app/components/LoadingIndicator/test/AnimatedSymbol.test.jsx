import React from "react";
import { shallow, mount } from "enzyme";

import * as WindowContext from "../../WindowContext";
import AnimatedSymbol from "../AnimatedSymbol";

const createWindow = (
    setTimeout = jest.fn(cb => {
        cb();
        return "some-id";
    }),
    clearTimeout = jest.fn()
) => ({
    setTimeout,
    clearTimeout,
});

const defaultProps = {
    basis: 100,
    offsetTop: 20,
};

const render = (overwriteProps, window, renderMethod = mount) => {
    jest.spyOn(WindowContext, "useWindowContext").mockImplementation(
        () => window
    );
    return renderMethod(
        <AnimatedSymbol {...defaultProps} {...overwriteProps}>
            some-symbol
        </AnimatedSymbol>
    );
};

describe("AnimatedSymbol", () => {
    it("should render children with correct initial styles", () => {
        const component = render({}, undefined, shallow);
        expect(component).toMatchSnapshot();
    });

    it("should set different opacity after random timeout", () => {
        const window = createWindow();
        const randomTimeout = 1;
        const getRandomInt = jest
            .fn()
            .mockReturnValueOnce(randomTimeout)
            .mockReturnValue(501);
        const component = render({ getRandomInt }, window);

        const styledSymbol = component.find("StyledSymbol");

        expect(styledSymbol.prop("opacity")).toEqual(1);
        expect(window.setTimeout.mock.calls[0][1]).toBe(randomTimeout);
    });

    it("should reset opacity after second random timeout", () => {
        const window = createWindow();
        const randomTimeout = 1;
        const getRandomInt = jest
            .fn()
            .mockReturnValueOnce(randomTimeout)
            .mockReturnValue(501);
        const component = render({ getRandomInt }, window);

        component.update();

        const styledSymbol = component.find("StyledSymbol");

        expect(styledSymbol.prop("opacity")).toEqual(1);
        expect(window.setTimeout.mock.calls[1][1]).toBe(501);

        expect(window.setTimeout.mock.calls.length).toEqual(2);
        expect(getRandomInt.mock.calls).toEqual([[100], [899]]);
    });

    it("call get random int with correct upper bounds", () => {
        const window = createWindow();
        const randomTimeout = 1;
        const getRandomInt = jest
            .fn()
            .mockReturnValueOnce(randomTimeout)
            .mockReturnValue(501);
        const staticClearanceValue = 100;
        const staticIntervalMinusAnimationTime = 899;
        render({ getRandomInt }, window);

        expect(getRandomInt.mock.calls).toEqual([
            [staticClearanceValue],
            [staticIntervalMinusAnimationTime],
        ]);
    });

    it("should blur again when children are replaced", () => {
        const window = createWindow();
        const randomTimeout = 1;
        const getRandomInt = jest.fn(() => randomTimeout);
        const component = render({ getRandomInt }, window);

        component.setProps({ children: "other-children" });

        expect(window.setTimeout.mock.calls.length).toEqual(4);
    });

    it("should clear timeouts on unmount", () => {
        const window = createWindow();
        window.setTimeout
            .mockImplementationOnce(cb => {
                cb();
                return "id-1";
            })
            .mockReturnValue("id-2");
        const component = render({}, window);

        component.unmount();

        expect(window.clearTimeout.mock.calls.length).toEqual(2);
        expect(window.clearTimeout.mock.calls[0][0]).toEqual("id-1");
        expect(window.clearTimeout.mock.calls[1][0]).toEqual("id-2");
    });
});
