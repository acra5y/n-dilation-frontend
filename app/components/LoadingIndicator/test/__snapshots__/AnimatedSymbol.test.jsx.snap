// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimatedSymbol should render children with correct initial styles 1`] = `
<StyledSymbol
  basis={100}
  offsetTop={20}
  opacity={0}
>
  some-symbol
</StyledSymbol>
`;
